src/osdc/Journaler.cc

```diff
@@ -493,7 +493,13 @@ void Journaler::_finish_write_head(int r, Header &wrote,
   }
   ceph_assert(!readonly);
   ldout(cct, 10) << "_finish_write_head " << wrote << dendl;
-  last_committed = wrote;
+  if (wrote.write_pos < last_committed.write_pos ||
+      wrote.expire_pos < last_committed.expire_pos ||
+      wrote.trimmed_pos < last_committed.trimmed_pos) {
+    lderr(cct) << __func__ << ": not updating last_committed" << dendl;
```

Comment:
I'm wary of just skipping the header update. Any objection to an actual assertion?

Reply:
Fair enough - I'll add the assert.

Reply:
@batrick why are you wary of skipping the header update here ?

Reply:
ping @mchangir @batrick

Reply:
> @batrick why are you wary of skipping the header update here ?

Because the Journaler had previously, successfully written a header with `last_committed` beyond the header write that just succeeded. This may indicate that we have clobbered the newest update to `head`. We cannot happily continue after having done so.

Looking at this code again, there's little to prevent this kind of error. It really should be harnessing the smarts of the OSD by encoding a sequence code -- the current write_pos --  in the write (probably stored in the omap) which fails the op if the write's sequence code is behind what the OSD has.

Reply:
> Looking at this code again, there's little to prevent this kind of error. It really should be harnessing the smarts of the OSD by encoding a sequence code -- the current write_pos -- in the write (probably stored in the omap) which fails the op if the write's sequence code is behind what the OSD has.

I agree and IMO this is something we would want to add at least when flushing journal header.